design: thin pool (#143)
red-hat-storage/lvm-operator: doc/design/thin_pool.md

# LVMO: Thin provisioning

## Summary

> Allow LVM operator to create thin-pools along with the volume groups.

Reviewer: ```suggestion
LVM thin provisioning allows creation of volumes whose combined size is greater than that of the available storage. 
Advantages: This leads to better storage utilization etc etc
Disadvantages : etc etc

LVM does this by allocating the blocks in a thin LV from a special "thin pool LV". A thin pool LV must be created before thin LVs can be created within it.

The LVMO will create a thin-pool LV in the volume group in order to create thinly provisioned volumes.
```

## Proposal:
- The `deviceClass` API in the `LVMClusterSpec` will contain the mapping between a volume group and a thin-pool.
- One volume group will be mapped to a single thin pool.
- User should be able to configure the thin-pool size based on percentage of the available volume group size.

> - Default chunk size of the thin pool should be 512 kiB

Reviewer: ```suggestion
- Default chunk size of the thin pool will be 512 kiB
```

+	// for creating the thin pool.
+	// +kubebuilder:validation:default=75
+	// +kubebuilder:validation:Minimum=10
	// +kubebuilder:validation:Maximum=90
+	SizePercent int `json:"sizePercent,omitempty"`

+	// OverProvisionRatio represents the ratio of overprovision that can
+	// be allowed on thin pools
+	OverprovisionRatio int `json:"overprovisionRatio,omitempty"`
}

type DeviceClass struct {
    Name string `json:"name,omitempty"`

    DeviceSelector *DeviceSelector `json:"deviceSelector,omitempty"`
    NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`

+   // ThinPoolConfig contains configurations for the thin-pool
+   ThinPoolConfig  ThinPoolConfig `json:"thinPoolConfig,omitempty"`
}
```

- Following new fields will added to `DeviceClass` API
    - **ThinPoolConfig** API contains information related to a thin pool.These configuration options are:
        - **Name**: Name of the thin-pool
        - **SizePercent**: Size of the thin pool to be created with respect to available free space in the volume group. It represents percentage value and not absolute size values. Size value should range between 10-90. It defaults to 75 if no value is provided.

>         - **OverprovisionRatio**: OverprovisionRatio represents the ratio of overprovision that can be allowed on thin pools.

Reviewer: ```suggestion
        - **OverprovisionRatio**: The factor by which additional storage can be provisioned compared to the available storage in the thin pool.
```

- `LVMVolumeGroup` API changes:

``` go=
type LVMVolumeGroupSpec struct {
    // DeviceSelector is a set of rules that should match for a device to be

>     // included in this TopoLVMCluster

Reviewer: ```suggestion
    // included in this VolumeGroup
```

	// +optional
    DeviceSelector *DeviceSelector `json:"deviceSelector,omitempty"`

    // NodeSelector chooses nodes
    // +optional
    NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`

+   // ThinPoolConfig contains configurations for the thin-pool
+   ThinPoolConfig ThinPoolConfig `json:"thinPoolConfig,omitempty"`
}
```

### VolumeGroup Manager
- Volume Group Manager is responsible for creating the thin-pools after creating the volume group.
- Command used for creating the thin-pool:
    ```
    lvcreate -L <Size>%FREE -c <Chunk size> -T <vg_name>/<thin-pool-name>
    ```
    where:
    - Size is `LVMClusterSpec.Storage.DeviceClass.ThinPoolConfig.SizePercent`
    - chunk size is 512KiB, which is the default.

> - Manager should also update the `lvmd.yaml` file to map volume group and its thin-pool

Reviewer: ```suggestion
- VG manager will also update the `lvmd.yaml` file to map volume group and its thin-pool to the topolvm device class
```

- Sample `lvmd.yaml` config file
``` yaml=
device-classes:
  - name: ssd-thin
    volume-group: myvg1
    spare-gb: 10
    type: thin
    thin-pool-config:
        name: pool0
        overprovision-ratio: 50.0
```

### Monitoring and Alerts
- Available thin-pool size (both data and metadata) should be provided by topolvm as prometheus metrics.
- Threshold limits for the thin-pool should be provide as static values in the prometheus yaml file.
- If available size of data or metadata for a particular thin-pool crosses the threshold, then appropriate alerts should be triggered.

### Open questions
- What should be the chunk size of the thin-pools?
    - Use default size a 512 kiB for now.

> - Should LVMO notify changes of thin-pool if it happens to topolvm?

Reviewer: Please clarify this point. What changes are you referring to ?

Reviewer 2: - this question can be dropped, topolvm by default updates annotations in 10 min intervals if no ops performed in that time space

- How should vg-manager handle expansion of thin-pool in near future?

> - After firing alerts should LVMO take any pre-cautionary steps to make thin-lvs occupying whole thin-pool without effecting snapshots

Reviewer: Please clarify this point as it is unclear.

Reviewer 2: - the original intention was, when thinpool gets fulled snapshots are dropped and as I wasn't able to recall where I read that this question can be dropped as well.

Author: dropped all the unanswered questions like suggested.

# LVMO: Thin provisioning

## Summary
- LVM thin provisioning allows creation of volumes whose combined size is greater than that of the available storage.

**Advantages**:
- Storage space can be used more effectively. More users can be accommodated for the same amount of storage space when compared to thick provisioning. This significantly reduces upfront hardware cost for the storage admins.
- Faster clones and snapshots

**Disadvantages** :
- Reduced performance when compared to thick volumes.
- Over-allocation of the space. This can be mitigated by better monitoring of the disk usage.

LVM does this by allocating the blocks in a thin LV from a special "thin pool LV". A thin pool LV must be created before thin LVs can be created within it.

The LVMO will create a thin-pool LV in the volume group in order to create thinly provisioned volumes.

## Proposal:

> - The `deviceClass` API in the `LVMClusterSpec` will contain the mapping between a volume group and a thin-pool.

Reviewer 2: ```suggestion
- The `deviceClass` API in the `LVMClusterSpec` will contain the mapping between a device-class and a thin-pool in volume group.
```

> - One volume group will be mapped to a single thin pool.

Reviewer 2: ```suggestion
- One device-class will be mapped to a single thin pool.
```

- User should be able to configure the thin-pool size based on percentage of the available volume group size.
- Default chunk size of the thin pool will be 512 kiB
- `lvmd.yaml` config file should be updated with the device class, volume group and thin-pool mapping.
- Alerts should be triggered if a thin-pool `data` or `metadata` available size crosses a predefined threshold limit.

## Design Details
### API changes:

- `LVMClusterSpec.Storage.DeviceClass.ThinPoolConfig` will have the mapping between device class, volume group and the thin-pool.
- One DeviceClass can be mapped to only one thin-pool.

- `LVMCluster` API changes
```go=
+ type ThinPoolConfig struct{
+    // Name of the thin pool to be creatd
+	Name string `json:"name,omitempty"`

+   // SizePercent represents percentage of remaining space in the volume group that should be used
+	// for creating the thin pool.
+	// +kubebuilder:validation:default=75
+	// +kubebuilder:validation:Minimum=10
	// +kubebuilder:validation:Maximum=90
+	SizePercent int `json:"sizePercent,omitempty"`

+	// OverProvisionRatio represents the ratio of overprovision that can
+	// be allowed on thin pools

> +	OverprovisionRatio int `json:"overprovisionRatio,omitempty"`

Reviewer 2: - this is still in question, it'll definitely be `int` need to know whether it represents percentage or multiplier
- either of the case, better to have `+kubebuilder:validation:Minimum`

}

type DeviceClass struct {
    Name string `json:"name,omitempty"`

    DeviceSelector *DeviceSelector `json:"deviceSelector,omitempty"`
    NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`

+   // ThinPoolConfig contains configurations for the thin-pool

> +   ThinPoolConfig  ThinPoolConfig `json:"thinPoolConfig,omitempty"`

Reviewer 2: - maybe use a pointer which helps in checking against `nil` for absence and drop the omitemtpy as well.

}
```

- Following new fields will added to `DeviceClass` API
    - **ThinPoolConfig** API contains information related to a thin pool.These configuration options are:
        - **Name**: Name of the thin-pool
        - **SizePercent**: Size of the thin pool to be created with respect to available free space in the volume group. It represents percentage value and not absolute size values. Size value should range between 10-90. It defaults to 75 if no value is provided.
        - **OverprovisionRatio**: The factor by which additional storage can be provisioned compared to the available storage in the thin pool.

- `LVMVolumeGroup` API changes:

``` go=
type LVMVolumeGroupSpec struct {
    // DeviceSelector is a set of rules that should match for a device to be
    // included in this VolumeGroup
	// +optional
    DeviceSelector *DeviceSelector `json:"deviceSelector,omitempty"`

    // NodeSelector chooses nodes
    // +optional
    NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`

+   // ThinPoolConfig contains configurations for the thin-pool

> +   ThinPoolConfig ThinPoolConfig `json:"thinPoolConfig,omitempty"`

Reviewer 2: - same as above

- Command used for creating the thin-pool:
    ```
    lvcreate -L <Size>%FREE -c <Chunk size> -T <vg_name>/<thin-pool-name>
    ```
    where:
    - Size is `LVMClusterSpec.Storage.DeviceClass.ThinPoolConfig.SizePercent`
    - chunk size is 512KiB, which is the default.

- VG manager will also update the `lvmd.yaml` file to map volume group and its thin-pool to the topolvm device class.
- Sample `lvmd.yaml` config file
``` yaml=
device-classes:
  - name: ssd-thin
    volume-group: myvg1
    spare-gb: 10
    type: thin
    thin-pool-config:
        name: pool0
        overprovision-ratio: 50.0
```

### Monitoring and Alerts
- Available thin-pool size (both data and metadata) should be provided by topolvm as prometheus metrics.
- Threshold limits for the thin-pool should be provide as static values in the prometheus yaml file.
- If available size of data or metadata for a particular thin-pool crosses the threshold, then appropriate alerts should be triggered.

### Open questions
- What should be the chunk size of the thin-pools?
    - Use default size a 512 kiB for now.
- Should LVMO notify changes of thin-pool if it happens to topolvm?

> - How should vg-manager handle expansion of thin-pool in near future?

Reviewer 2: - as we are not tracking state of VG or thinpool and lvmo doesn't get effected by these changes this question can be dropped

- After firing alerts should LVMO take any pre-cautionary steps to make thin-lvs occupying whole thin-pool without effecting snapshots

> - Do we want to leverage dm-eventd for auto-expansion of thinpool, how could Topolvm monitoring and dm-eventd be made to work in tandem

Reviewer 2: - as we'll be using topolvm solely for monitoring, no need for dm-eventd for auto-expansion, this question can be dropped

    // NodeSelector chooses nodes
    // +optional
    NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`

+   // ThinPoolConfig contains configurations for the thin-pool
+   // +kubebuilder:validation:Required
+   // +required
+   ThinPoolConfig *ThinPoolConfig `json:"thinPoolConfig,omitempty"`
}
```

### VolumeGroup Manager
- Volume Group Manager is responsible for creating the thin-pools after creating the volume group.
- Command used for creating the thin-pool:
    ```
    lvcreate -L <Size>%FREE -c <Chunk size> -T <vg_name>/<thin-pool-name>
    ```
    where:
    - Size is `LVMClusterSpec.Storage.DeviceClass.ThinPoolConfig.SizePercent`
    - chunk size is 512KiB, which is the default.

- VG manager will also update the `lvmd.yaml` file to map volume group and its thin-pool to the topolvm device class.
- Sample `lvmd.yaml` config file
``` yaml=
device-classes:
  - name: ssd-thin
    volume-group: myvg1
    spare-gb: 10
    type: thin
    thin-pool-config:
        name: pool0
        overprovision-ratio: 50.0
```

### Monitoring and Alerts
- Available thin-pool size (both data and metadata) should be provided by topolvm as prometheus metrics.

> - Threshold limits for the thin-pool should be provide as static values in the prometheus yaml file.

Reviewer: ```suggestion
- Threshold limits for the thin-pool should be provide as static values in the PrometheusRule.
```

> - If available size of data or metadata for a particular thin-pool crosses the threshold, then appropriate alerts should be triggered.

Reviewer: ```suggestion
- If used size of data or metadata for a particular thin-pool crosses the threshold, then appropriate alerts should be triggered.
```
